chainer/iterators/multiprocess_iterator.py

```diff
@@ -41,45 +51,93 @@ class MultiprocessIterator(iterator.Iterator):
 
     """
 
-    _last_signal = object()
+    _interruption_testing = False  # for testing
 
     def __init__(self, dataset, batch_size, repeat=True, shuffle=True,
                  n_processes=None, n_prefetch=1, shared_mem=None):
         self.dataset = dataset
         self.batch_size = batch_size
-        self._repeat = repeat
-        self._shuffle = shuffle
-        self._prefetch_order = None  # used at the end of each epoch
+        self.repeat = repeat
+        self.shuffle = shuffle
 
         self.n_processes = n_processes or multiprocessing.cpu_count()
         self.n_prefetch = max(n_prefetch, 1)
-        self._shared_mem_size = shared_mem
+        self.shared_mem = shared_mem
 
-        self._finalized = None
+        self._finalized = False
 
+        self._comm = _Communicator(self.n_prefetch)
         self.reset()
 
-    def __del__(self):
-        self.finalize()
+        self._prefetch_loop = _PrefetchLoop(
+            self.dataset, self.batch_size, self.repeat, self.shuffle,
+            self.n_processes, self.n_prefetch, self.shared_mem, self._comm,
+            self.__class__._interruption_testing)
+        # defer launching prefetch thread until creating the worker pool,
+        # not to leave a background thread in forked processes.
+        self._thread = None
 
     def __next__(self):
-        if not self._repeat and self.epoch > 0:
+        measure_mode = False
+        if self._thread is None:
+            if self._prefetch_loop.measure_required():
+                measure_mode = True
+                batch, prefetch_state = self._prefetch_loop.measure()
+            self._thread = self._prefetch_loop.launch_thread()
+            del self._prefetch_loop
+
+        if not measure_mode:
+            batch, prefetch_state = self._comm.get()
+
+        (self.current_position, self.epoch, self.is_new_epoch,
+            self._previous_epoch_detail, self._order) = prefetch_state
+        if batch is None:
             raise StopIteration
+        else:
+            return batch
 
-        self._previous_epoch_detail = self.epoch_detail
+    next = __next__
 
-        self.is_new_epoch = False
-        if self._finalized is None:
-            self._init()  # start workers
-            # load for the first iteration
-            for _ in six.moves.range(self.n_prefetch):
-                self._invoke_prefetch()
+    def __del__(self):
+        try:
```

Comment:
Why do you use `try`?

Reply:
OK, I'll change the code to assign `None` for avoiding `try`.
I'll add comments about that attributes may not be initialized at `__del__`, too.